---
title: "README"
author: "Robert Reid"
date: "`r Sys.Date()`"
output: {md_document: {variant: 'markdown_github'}}
bibliography: references/ref.bib
---

```{r setup, include=FALSE}
library(rmarkdown)
library(pandoc)
knitr::opts_chunk$set(echo = TRUE)
```

## CAS9 Volcano plot

This R project file contains code and data used to produce figures 4A and S4C
in the manuscript _Leading and lagging strand fork collapses differ in structure
and requirements for repair_ by Michael T Kimble, Aakanksha Sane,
Robert JD Reid, Meret Arter, Matthew J Johnson, Rodney Rothstein, Scott Keeney
and Lorraine S Symington. In revision for _Molecular Cell_.

A mutant version of _CAS9_ was used to make single strand breaks in genomic DNA using
RNA guides targeting sites on yeast chromosome VI. A CAS9 and guide
RNA expression plasmid was transferred into the approximately 4800 strains
of a yeast gene disruption library using the selective ploidy ablation (SPA)
protocol [@reid_selective_2011]. CAS9 expression was induced in these mutant
strains, growth was determined by colony size using the _screenmill_ R package [@reid_screenmill]
and growth of each strain was compared to a scrambled guide control or an empty vector control.

This project imports processed data from _screenmill_ output, combines data from 
two runs of the SPA screen, and calculates a $t$-statistic and $p$-value for the
growth comparison. These are plotted in a "volcano" plot which is normalized
growth difference by $p$-value shown below. Note that _screenmill_ requires a version
less than 4.2 due to a breaking change in one of the packages it relies on.



```{r, out.width='50%', fig.align='center',echo=FALSE}
knitr::include_graphics("figures/dif_scr.png")
```

## References
